doc/io_wip.md

```diff
@@ -0,0 +1,206 @@
+`Pull()` method that can be called to either peek at or extract a chunk
+of data. The caller passes in a buffer. The `Source` chooses to either
+copy it's data into that buffer, or return it's own buffer. Whatever it
+chooses, the data is returned to the caller by invoking the pull_cb.
+The `Source` may provide it's own callback (the push_cb) at that point
+to be notified when the caller is done processing the data.
+
+The Pull operation may be sync or async. The caller can require sync.
+
+The Pull operation may be a peek (e.g. non-committing read)
+
+The Source may buffer data if necessary.
+
+### Example 1: Simple, synchronous, strict length read
+
+In the following example, the caller is telling the source that it:
+
+* Wants the pull to complete synchronously
+* Wants the Source to copy the data into the provided buffer.
+
+The source will return an error if it is not able to meet these restrictions.
+
+```c
+Source* mySource = GetSourceSomehow();
+io_buf_t buffer;
+buffer.base = Malloc<uint8_t*>(100);
+buffer.len = 100;
+
+io_pull_t pull;
+
+int status = mySource->Pull(&pull, &buffer.len, &buffer, 1, IO_PULL_FLAG_SYNC);
+
+CHECK(status & IO_PULL_STATUS_SYNC);
+if (status & IO_PULL_STATUS_ERROR) {
+  int err = io_get_error(status);
+  // do something with the error
+  return;
+}
+
+// the data should be in buffer
+```
+
+### Example 2: Peek first, then read sync
+
+```c
+Source* mySource = GetSourceSomehow();
+
+size_t length = 100;
+
+io_pull_t pull;
+
+// Peek to get the length
+int status = mySource->Pull(&pull, &len, nullptr, 0,
+                            IO_PULL_FLAG_SYNC |
+                            IO_PULL_FLAG_PEEK);
+
+CHECK(status & IO_PULL_STATUS_SYNC);
+if (status & IO_PULL_STATUS_ERROR) {
+  int err = io_get_error(status);
+  // do something with the error
+  return;
+}
+
+io_buf_t buffer;
+buffer.base = Malloc<uint8_t*>(length);
+buffer.len = length;
+
+status = mySource->Pull(&pull, &length, &buffer, 1, IO_PULL_FLAG_SYNC);
+
+CHECK(status & IO_PULL_STATUS_SYNC);
+if (status & IO_PULL_STATUS_ERROR) {
+  int err = io_get_error(status);
+  // do something with the error
+  return;
+}
+
+// the data should be in buffer
+```
+
+### Example 3: Async pull, data copied into provided buffer
+
+```c
+Source* mySource = GetSourceSomehow();
+
+io_buf_t buffer;
+buffer.base = Malloc<uint8_t*>(100);
+buffer.len = length;
+
+auto AfterPull = [](io_pull_t* handle,
+                    int status,
+                    io_buf_t* bufs,
+                    size_t count,
+                    size_t length,
+                    io_push_t* source) {
+  if (status & IO_PULL_STATUS_ERROR) {
+    int error = io_get_error(status);
+    // handle the error;
+    return;
+  }
+
+  // data should be in bufs
+};
+
+io_pull_t pull;
+io_pull_set_pull_cb(&pull, AfterPull);
+
+int status = mySource->Pull(&pull, &buffer.len, &buffer, 1,
+                            IO_PULL_FLAG_MUST_COPY);
```

Comment:
Is the idea here that `pull` is copied into memory managed by the source object?

Reply:
`pull` is just an async handle, really. Used to maintain context in the callback.

The `IO_PULL_FLAG_MUST_COPY` tells the `Source` that it *must* copy it's data into the provided `buffer` rather than providing pointers to it's own buffers in the callback.